File: android/app/src/main/kotlin/com/siezagym/app/Features/Profile/ProfileScreen.kt
package com.siezagym.app.Features.Profile

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil3.compose.AsyncImage
import com.siezagym.app.DesignSystem.CapsuleShape
import com.siezagym.app.DesignSystem.D2Theme
import com.siezagym.app.DesignSystem.FilaLista
import com.siezagym.app.DesignSystem.GhostButton
import com.siezagym.app.DesignSystem.GlassCard
import com.siezagym.app.DesignSystem.LocalD2Theme
import com.siezagym.app.DesignSystem.LocalThemeStore
import com.siezagym.app.DesignSystem.PanelLista
import com.siezagym.app.DesignSystem.SectionLabel
import com.siezagym.app.DesignSystem.SolidButton
import com.siezagym.app.DesignSystem.StatsCard
import com.siezagym.app.DesignSystem.TemasTokens
import com.siezagym.app.DesignSystem.Pantalla
import com.siezagym.app.Features.Shared.NavInset
import com.siezagym.app.Models.ExperienceLevel
import com.siezagym.app.Models.Sex
import com.siezagym.app.Models.UserProfile
import com.siezagym.app.R
import com.siezagym.app.Services.AuthService
import com.siezagym.app.Services.GymStore
import androidx.compose.runtime.collectAsState
import kotlinx.coroutines.launch

/** Perfil, igual que `/perfil` en la web: la tarjeta de identidad, tres
 *  números, los datos de la cuenta y la configuración con el tema adentro. */
@Composable
fun ProfileScreen(store: GymStore, auth: AuthService) {
    val tema = LocalD2Theme.current
    val perfil by store.profile.collectAsState()
    val sesiones by store.sessions.collectAsState()
    val authError by auth.errorMessage.collectAsState()
    val scope = rememberCoroutineScope()

    var peso by remember { mutableStateOf("") }
    var altura by remember { mutableStateOf("") }
    var meta by remember { mutableStateOf("") }
    var sexo by remember { mutableStateOf<Sex?>(null) }
    var nivel by remember { mutableStateOf<ExperienceLevel?>(null) }
    var guardado by remember { mutableStateOf(false) }
    var cargado by remember { mutableStateOf(false) }

    // Solo la primera vez: después el texto es del usuario, no se pisa.
    LaunchedEffect(perfil) {
        if (!cargado) {
            perfil?.let { p ->
                peso = p.bodyWeightKg?.let(::numeroTexto) ?: ""
                altura = p.heightCm?.let(::numeroTexto) ?: ""
                meta = p.weeklyCalorieGoalKcal?.let(::numeroTexto) ?: ""
                sexo = p.sex
                nivel = p.experienceLevel
                cargado = true
            }
        }
    }

    val totalSeries = sesiones.sumOf { it.totalSetsCompleted }

    // Un cambio vuelve a desactivar el "Listo, guardado." del Swift.
    fun marcarCambio() {
        if (guardado) guardado = false
    }

    Pantalla(titulo = "Perfil") {
        Spacer(Modifier.height(20.dp))
        Identidad(perfil)

        Spacer(Modifier.height(14.dp))
        StatsCard(datos = listOf(
            "${sesiones.size}" to plural(sesiones.size, "entrenamiento", "entrenamientos"),
            "$totalSeries" to plural(totalSeries, "serie", "series"),
            "${store.streak}" to plural(store.streak, "día seguido", "días seguidos"),
        ))

        SectionLabel("Tus datos", modifier = Modifier.padding(top = 24.dp, bottom = 10.dp))
        GlassCard(modifier = Modifier.fillMaxWidth(), paddingInterno = 16.dp) {
            Column {
                Opciones(
                    rotulo = "Sexo",
                    todas = Sex.entries,
                    seleccion = sexo,
                    etiqueta = { it.label },
                    onElegir = { opcion ->
                        sexo = if (sexo == opcion) null else opcion
                        marcarCambio()
                    },
                )
                Spacer(Modifier.height(16.dp))
                Opciones(
                    rotulo = "Experiencia",
                    todas = ExperienceLevel.entries,
                    seleccion = nivel,
                    etiqueta = { it.label },
                    onElegir = { opcion ->
                        nivel = if (nivel == opcion) null else opcion
                        marcarCambio()
                    },
                )
                Spacer(Modifier.height(16.dp))
                Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                    Campo("Peso (kg)", peso, Modifier.weight(1f)) {
                        peso = it
                        marcarCambio()
                    }
                    Campo("Altura (cm)", altura, Modifier.weight(1f)) {
                        altura = it
                        marcarCambio()
                    }
                }
                Spacer(Modifier.height(16.dp))
                Campo("Meta semanal (kcal)", meta, Modifier.fillMaxWidth()) {
                    meta = it
                    marcarCambio()
                }
                Spacer(Modifier.height(16.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    SolidButton(
                        texto = "Guardar",
                        onClick = {
                            scope.launch {
                                store.updateProfile(
                                    mapOf(
                                        // Vacío se guarda como null y no como cero: son cosas distintas.
                                        "bodyWeightKg" to aNumero(peso),
                                        "heightCm" to aNumero(altura),
                                        "weeklyCalorieGoalKcal" to aNumero(meta),
                                        "sex" to sexo?.raw,
                                        "experienceLevel" to nivel?.raw,
                                    ),
                                )
                                guardado = true
                            }
                        },
                        expands = false,
                    )
                    if (guardado) {
                        Spacer(Modifier.width(12.dp))
                        Text("Listo, guardado.", fontSize = 12.sp, color = tema.texto2)
                    }
                }
                Spacer(Modifier.height(16.dp))
                // El peso no es decorativo: la portada lo usa para estimar las
                // calorías de la semana.
                Text(
                    "El peso se usa para estimar las calorías de cada sesión.",
                    fontSize = 11.sp,
                    color = tema.texto3,
                )
            }
        }

        SectionLabel("Configuración", modifier = Modifier.padding(top = 24.dp, bottom = 10.dp))
        Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
            GlassCard(modifier = Modifier.fillMaxWidth(), paddingInterno = 16.dp) {
                Column {
                    Text("Tema", fontSize = 14.sp, fontWeight = FontWeight.Medium, color = tema.texto)
                    Spacer(Modifier.height(4.dp))
                    Text(
                        "El fondo de la app. Se guarda en este teléfono.",
                        fontSize = 11.sp,
                        color = tema.texto2,
                    )
                    Spacer(Modifier.height(12.dp))
                    SelectorTemas()
                }
            }
            PanelLista(modifier = Modifier.fillMaxWidth()) {
                FilaLista(
                    nombre = "Entrenador",
                    detalle = "Quien te asigna rutinas",
                    valor = if (perfil?.isCoach == true) "Sos entrenador" else "Sin vincular",
                    chevron = false,
                )
            }
        }

        GhostButton(
            texto = "Cerrar sesión",
            onClick = { auth.signOut() },
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 24.dp),
        )

        authError?.let { mensaje ->
            Spacer(Modifier.height(12.dp))
            Text(mensaje, fontSize = 13.sp, color = tema.texto2)
        }

        Spacer(Modifier.height(NavInset.bottom.dp))
    }
}

// MARK: - Identidad

@Composable
private fun Identidad(perfil: UserProfile?) {
    val tema = LocalD2Theme.current
    GlassCard(modifier = Modifier.fillMaxWidth(), paddingInterno = 16.dp, radius = 26.dp) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Box(
                Modifier
                    .size(58.dp)
                    .clip(CircleShape)
                    .background(tema.vidrio(2))
                    .border(1.dp, tema.borde, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    perfil?.initial ?: "T",
                    fontSize = 24.sp,
                    fontWeight = FontWeight.Medium,
                    color = tema.texto,
                )
                // La inicial queda atrás y actúa de placeholder como en iOS.
                if (!perfil?.photoUrl.isNullOrBlank()) {
                    AsyncImage(
                        model = perfil?.photoUrl,
                        contentDescription = null,
                        contentScale = ContentScale.Crop,
                        modifier = Modifier.fillMaxSize(),
                    )
                }
            }
            Spacer(Modifier.width(14.dp))
            Column(Modifier.weight(1f)) {
                Text(
                    perfil?.displayName ?: "Sin nombre",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Medium,
                    color = tema.texto,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                perfil?.email?.let { email ->
                    Text(
                        email,
                        fontSize = 12.sp,
                        color = tema.texto2,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
                Text(
                    if (perfil?.isCoach == true) "Entrenador" else "Atleta",
                    fontSize = 11.sp,
                    color = tema.texto3,
                )
            }
        }
    }
}

// MARK: - Datos

/** Opciones en línea. Volver a tocar la elegida la desmarca: nada es
 *  obligatorio. */
@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun <T : Any> Opciones(
    rotulo: String,
    todas: List<T>,
    seleccion: T?,
    etiqueta: (T) -> String,
    onElegir: (T) -> Unit,
) {
    val tema = LocalD2Theme.current
    Column {
        Text(rotulo, fontSize = 11.sp, color = tema.texto2)
        Spacer(Modifier.height(8.dp))
        FlowRow(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            todas.forEach { opcion ->
                val activa = seleccion == opcion
                Box(
                    Modifier
                        .clip(CapsuleShape)
                        .background(if (activa) tema.solido else tema.vidrio(1))
                        .border(
                            width = 1.dp,
                            color = if (activa) Color.Transparent else tema.borde,
                            shape = CapsuleShape,
                        )
                        .clickable { onElegir(opcion) }
                        .heightIn(min = 38.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        etiqueta(opcion),
                        fontSize = 13.sp,
                        color = if (activa) tema.sobreSolido else tema.texto2,
                        textAlign = TextAlign.Center,
                        modifier = Modifier.padding(horizontal = 14.dp, vertical = 6.dp),
                    )
                }
            }
        }
    }
}

/** Campo numérico en vidrio: rótulo chico arriba y el número centrado. */
@Composable
private fun Campo(
    rotulo: String,
    valor: String,
    modifier: Modifier = Modifier,
    onChange: (String) -> Unit,
) {
    val tema = LocalD2Theme.current
    Column {
        Text(rotulo, fontSize = 11.sp, color = tema.texto2)
        Spacer(Modifier.height(8.dp))
        TextField(
            value = valor,
            onValueChange = onChange,
            singleLine = true,
            placeholder = { Text("—", color = tema.texto3) },
            shape = RoundedCornerShape(14.dp),
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
            textStyle = androidx.compose.ui.text.TextStyle(
                color = tema.texto,
                fontSize = 16.sp,
                textAlign = TextAlign.Center,
            ),
            colors = TextFieldDefaults.colors(
                focusedContainerColor = tema.vidrio(1),
                unfocusedContainerColor = tema.vidrio(1),
                disabledContainerColor = tema.vidrio(1),
                focusedIndicatorColor = Color.Transparent,
                unfocusedIndicatorColor = Color.Transparent,
                disabledIndicatorColor = Color.Transparent,
                cursorColor = tema.solido,
                focusedPlaceholderColor = tema.texto3,
                unfocusedPlaceholderColor = tema.texto3,
                disabledPlaceholderColor = tema.texto3,
            ),
            modifier = modifier
                .clip(RoundedCornerShape(14.dp))
                .border(1.dp, tema.borde, RoundedCornerShape(14.dp))
                .height(46.dp),
        )
    }
}

// MARK: - Configuración

/** La grilla de temas del selector, como la `LazyVGrid` de iOS: tres por fila
 *  y la elegida marcada con el vidrio. */
@Composable
private fun SelectorTemas() {
    val tema = LocalD2Theme.current
    val temas = LocalThemeStore.current
    val filas = TemasTokens.chunked(3)
    Column {
        filas.forEachIndexed { indice, fila ->
            Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                fila.forEach { opcion ->
                    val activa = opcion.id == temas.actual.id
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        modifier = Modifier
                            .weight(1f)
                            .clip(RoundedCornerShape(18.dp))
                            .background(if (activa) tema.vidrio(1) else Color.Transparent)
                            .border(
                                width = if (activa) 1.dp else 0.dp,
                                color = if (activa) tema.bordeFuerte else Color.Transparent,
                                shape = RoundedCornerShape(18.dp),
                            )
                            .clickable { temas.elegir(opcion.id) }
                            .padding(vertical = 10.dp),
                    ) {
                        MuestraTema(opcion)
                        Spacer(Modifier.height(8.dp))
                        Text(
                            opcion.nombre,
                            fontSize = 12.sp,
                            color = if (activa) tema.texto else tema.texto2,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                        )
                    }
                }
                repeat(3 - fila.size) { Spacer(Modifier.weight(1f)) }
            }
            if (indice < filas.lastIndex) {
                Spacer(Modifier.height(12.dp))
            }
        }
    }
}

/** El fondo real del tema en un círculo: la obra si la trae, o su degradado. */
@Composable
private fun MuestraTema(opcion: D2Theme) {
    val tema = LocalD2Theme.current
    Box(
        Modifier
            .size(46.dp)
            .clip(CircleShape)
            .background(
                Brush.linearGradient(
                    colorStops = *opcion.fondo.map { it.location to it.color }.toTypedArray(),
                    start = Offset(46f * opcion.fondoInicio.first, 46f * opcion.fondoInicio.second),
                    end = Offset(46f * opcion.fondoFin.first, 46f * opcion.fondoFin.second),
                ),
            )
            .border(1.dp, tema.bordeFuerte, CircleShape),
    ) {
        val obra = when (opcion.obra) {
            "pliegues" -> R.drawable.fondo_pliegues
            "electrico" -> R.drawable.fondo_electrico
            else -> null
        }
        if (obra != null) {
            Image(
                painter = painterResource(obra),
                contentDescription = null,
                contentScale = ContentScale.Crop,
                modifier = Modifier.fillMaxSize(),
            )
        }
    }
}

// MARK: - Utilidades

private fun plural(cantidad: Int, uno: String, varios: String): String =
    if (cantidad == 1) uno else varios

/** Número del perfil a texto, sin decimales cuando no los tiene. */
private fun numeroTexto(valor: Double): String =
    if (valor % 1.0 == 0.0) valor.toLong().toString() else valor.toString()

/** Lo que escribe el usuario a número. La coma decimal también se acepta. */
private fun aNumero(texto: String): Double? = texto.replace(',', '.').toDoubleOrNull()

File: ios/SiezaGym/Features/Profile/ProfileScreen.swift
import SwiftUI

/// Perfil, igual que `/perfil` en la web: la tarjeta de identidad, tres
/// números, los datos de la cuenta y la configuración con el tema adentro.
struct ProfileScreen: View {
    @Environment(\.tema) private var tema
    @Environment(ThemeStore.self) private var temas
    @Environment(AuthService.self) private var auth
    let store: GymStore

    @State private var peso = ""
    @State private var altura = ""
    @State private var meta = ""
    @State private var sexo: Sex?
    @State private var nivel: ExperienceLevel?
    @State private var guardadoEn: Date?
    @State private var cargado = false

    private var totalSeries: Int { store.sessions.reduce(0) { $0 + $1.totalSetsCompleted } }

    var body: some View {
        NavigationStack {
            Pantalla(titulo: "Perfil") {
                identidad.padding(.top, 20)

                StatsCard(datos: [
                    ("\(store.sessions.count)", store.sessions.count == 1 ? "entrenamiento" : "entrenamientos"),
                    ("\(totalSeries)", totalSeries == 1 ? "serie" : "series"),
                    ("\(store.streak)", store.streak == 1 ? "día seguido" : "días seguidos"),
                ])
                .padding(.top, 14)

                SectionLabel("Tus datos").padding(.top, 24).padding(.bottom, 10)
                datos

                SectionLabel("Configuración").padding(.top, 24).padding(.bottom, 10)
                configuracion

                Button("Cerrar sesión") { auth.signOut() }
                    .buttonStyle(GhostButtonStyle())
                    .frame(maxWidth: .infinity)
                    .padding(.top, 24)
            }
            .bottomNavInset()
            .task { if !cargado { llenarDesdePerfil() } }
            .animation(.smooth(duration: 0.25), value: guardadoEn)
        }
        // Peso, altura y meta usan teclado decimal, que no trae tecla de cerrar.
        .tecladoConBotonListo()
    }

    // MARK: - Identidad

    private var identidad: some View {
        GlassCard(padding: 16, radius: 26) {
            HStack(spacing: 14) {
                Group {
                    if let url = store.profile?.photoURL {
                        AsyncImage(url: url) { $0.resizable().aspectRatio(contentMode: .fill) } placeholder: { inicial }
                    } else {
                        inicial
                    }
                }
                .frame(width: 58, height: 58)
                .clipShape(.circle)
                .overlay { Circle().strokeBorder(tema.borde, lineWidth: 1) }

                VStack(alignment: .leading, spacing: 4) {
                    Text(store.profile?.displayName ?? "Sin nombre")
                        .font(.system(size: 18, weight: .medium))
                        .foregroundStyle(tema.texto)
                        .lineLimit(1)
                    if let email = store.profile?.email {
                        Text(email)
                            .font(.system(size: 12))
                            .foregroundStyle(tema.texto2)
                            .lineLimit(1)
                    }
                    Text(store.profile?.isCoach == true ? "Entrenador" : "Atleta")
                        .font(.system(size: 11))
                        .foregroundStyle(tema.texto3)
                }
                .frame(maxWidth: .infinity, alignment: .leading)
            }
        }
    }

    private var inicial: some View {
        ZStack {
            tema.vidrio(2)
            Text(store.profile?.initial ?? "T")
                .font(.system(size: 24, weight: .medium))
                .foregroundStyle(tema.texto)
        }
    }

    // MARK: - Datos

    private var datos: some View {
        GlassCard(padding: 16) {
            VStack(alignment: .leading, spacing: 16) {
                opciones("Sexo", Sex.allCases, seleccion: sexo) { sexo = sexo == $0 ? nil : $0 }
                opciones("Experiencia", ExperienceLevel.allCases, seleccion: nivel) { nivel = nivel == $0 ? nil : $0 }

                HStack(spacing: 12) {
                    campo("Peso (kg)", texto: $peso)
                    campo("Altura (cm)", texto: $altura)
                }
                campo("Meta semanal (kcal)", texto: $meta)

                HStack(spacing: 12) {
                    Button("Guardar") { Task { await guardar() } }
                        .buttonStyle(SolidButtonStyle(expands: false))
                    if guardadoEn != nil {
                        Text("Listo, guardado.")
                            .font(.system(size: 12))
                            .foregroundStyle(tema.texto2)
                    }
                }

                // El peso no es decorativo: la portada lo usa para estimar las
                // calorías de la semana.
                Text("El peso se usa para estimar las calorías de cada sesión.")
                    .font(.system(size: 11))
                    .foregroundStyle(tema.texto3)
            }
        }
    }

    private func campo(_ rotulo: String, texto: Binding<String>) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(rotulo).font(.system(size: 11)).foregroundStyle(tema.texto2)
            TextField("", text: texto, prompt: Text("—").foregroundStyle(tema.texto3))
                .keyboardType(.decimalPad)
                .foregroundStyle(tema.texto)
                .multilineTextAlignment(.center)
                .padding(.horizontal, 14)
                .frame(minHeight: 46)
                .background(tema.vidrio(1), in: .rect(cornerRadius: 14))
                .overlay { RoundedRectangle(cornerRadius: 14).strokeBorder(tema.borde, lineWidth: 1) }
                .onChange(of: texto.wrappedValue) { if guardadoEn != nil { guardadoEn = nil } }
        }
        .frame(maxWidth: .infinity)
    }

    /// Opciones en línea. Volver a tocar la elegida la desmarca: nada es
    /// obligatorio.
    private func opciones<T: Hashable & Etiquetable>(
        _ rotulo: String, _ todas: [T], seleccion: T?, elegir: @escaping (T) -> Void
    ) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(rotulo).font(.system(size: 11)).foregroundStyle(tema.texto2)
            FlowRow(spacing: 8) {
                ForEach(todas, id: \.self) { opcion in
                    let activa = seleccion == opcion
                    Button {
                        elegir(opcion)
                        if guardadoEn != nil { guardadoEn = nil }
                    } label: {
                        Text(opcion.label)
                            .font(.system(size: 13))
                            .foregroundStyle(activa ? tema.sobreSolido : tema.texto2)
                            .padding(.horizontal, 14)
                            .frame(minHeight: 38)
                            .background(activa ? tema.solido : tema.vidrio(1), in: .capsule)
                            .overlay { Capsule().strokeBorder(activa ? .clear : tema.borde, lineWidth: 1) }
                    }
                    .buttonStyle(.plain)
                }
            }
        }
    }

    // MARK: - Configuración

    private var configuracion: some View {
        VStack(spacing: 14) {
            GlassCard(padding: 16) {
                VStack(alignment: .leading, spacing: 12) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text("Tema").font(.system(size: 14, weight: .medium)).foregroundStyle(tema.texto)
                        Text("La apariencia de la app. Se guarda en este teléfono.")
                            .font(.system(size: 11)).foregroundStyle(tema.texto2)
                    }

                    LazyVGrid(columns: Array(repeating: GridItem(.flexible(), spacing: 10), count: 3), spacing: 12) {
                        ForEach(Theme.todos) { opcion in
                            Button { temas.actual = opcion } label: {
                                VStack(spacing: 8) {
                                    muestra(opcion)
                                    Text(opcion.nombre)
                                        .font(.system(size: 12)).lineLimit(1).minimumScaleFactor(0.8)
                                }
                                .foregroundStyle(opcion.id == tema.id ? tema.texto : tema.texto2)
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 10)
                                .background(opcion.id == tema.id ? tema.vidrio(1) : .clear, in: .rect(cornerRadius: 18))
                                .overlay {
                                    RoundedRectangle(cornerRadius: 18)
                                        .strokeBorder(opcion.id == tema.id ? tema.bordeFuerte : .clear, lineWidth: 1)
                                }
                                .contentShape(.rect)
                            }
                            .buttonStyle(.plain)
                            .accessibilityLabel(opcion.nombre)
                            .accessibilityAddTraits(opcion.id == tema.id ? [.isSelected] : [])
                        }
                    }
                }
            }
            .animation(.smooth(duration: 0.25), value: tema.id)

            PanelLista {
                FilaLista(nombre: "Entrenador",
                          detalle: "Quien te asigna rutinas",
                          valor: store.profile?.isCoach == true ? "Sos entrenador" : "Sin vincular",
                          chevron: false)
            }
        }
    }

    /// El fondo real del tema en un círculo: la obra si la trae, o su degradado.
    private func muestra(_ opcion: Theme) -> some View {
        ZStack {
            LinearGradient(stops: opcion.fondo, startPoint: opcion.fondoInicio, endPoint: opcion.fondoFin)
            if let obra = opcion.obra {
                Image(obra).resizable().aspectRatio(contentMode: .fill)
            }
            if opcion.plano {
                Circle()
                    .fill(opcion.solido)
                    .frame(width: 16, height: 16)
            }
        }
        .frame(width: 46, height: 46)
        .clipShape(.circle)
        .overlay { Circle().strokeBorder(tema.bordeFuerte, lineWidth: 1) }
    }

    // MARK: - Datos del perfil

    private func llenarDesdePerfil() {
        guard let perfil = store.profile else { return }
        peso = perfil.bodyWeightKg.map { $0.formatted() } ?? ""
        altura = perfil.heightCm.map { $0.formatted() } ?? ""
        meta = perfil.weeklyCalorieGoalKcal.map { $0.formatted() } ?? ""
        sexo = perfil.sex
        nivel = perfil.experienceLevel
        cargado = true
    }

    private func guardar() async {
        var campos: [String: Any] = [:]
        // Vacío se guarda como null y no como cero: son cosas distintas.
        campos["bodyWeightKg"] = Double(peso.replacingOccurrences(of: ",", with: ".")) ?? NSNull()
        campos["heightCm"] = Double(altura.replacingOccurrences(of: ",", with: ".")) ?? NSNull()
        campos["weeklyCalorieGoalKcal"] = Double(meta.replacingOccurrences(of: ",", with: ".")) ?? NSNull()
        campos["sex"] = sexo?.rawValue ?? NSNull()
        campos["experienceLevel"] = nivel?.rawValue ?? NSNull()
        await store.updateProfile(campos)
        guardadoEn = Date()
    }
}

/// Lo que necesita `opciones` de cada opción: cómo se escribe.
protocol Etiquetable { var label: String { get } }
extension Sex: Etiquetable {}
extension ExperienceLevel: Etiquetable {}

/// Fila que envuelve: "Prefiero no decir" no entra en un tercio de pantalla.
struct FlowRow: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let ancho = proposal.width ?? .infinity
        var x: CGFloat = 0, y: CGFloat = 0, altoFila: CGFloat = 0
        for vista in subviews {
            let tamano = vista.sizeThatFits(.unspecified)
            if x + tamano.width > ancho, x > 0 {
                x = 0; y += altoFila + spacing; altoFila = 0
            }
            x += tamano.width + spacing
            altoFila = max(altoFila, tamano.height)
        }
        return CGSize(width: ancho, height: y + altoFila)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        var x = bounds.minX, y = bounds.minY, altoFila: CGFloat = 0
        for vista in subviews {
            let tamano = vista.sizeThatFits(.unspecified)
            if x + tamano.width > bounds.maxX, x > bounds.minX {
                x = bounds.minX; y += altoFila + spacing; altoFila = 0
            }
            vista.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(tamano))
            x += tamano.width + spacing
            altoFila = max(altoFila, tamano.height)
        }
    }
}
